### PR TITLE
ascanrulesAlpha: FP fixes for time based NoSQL

### PR DESCRIPTION
**Pull request by itsdivyanshjain**
Hey zap team, divyansh from astra security here. This PR contains fix for false positive which raises based on time-based nosql injection. Implemented solution is based on this PR:https://github.com/zaproxy/zap-extensions/pull/4173/files

### CODE REVIEW

**Comment by thc202** [comment]
The commit needs to be signed-off. (Squash the commits.)
The changelog should be updated.

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/timing/OnlineSimpleLinearRegression.java
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/timing/OnlineSimpleLinearRegression.java:36]: It might be a good time to move these to `commonlib`.

**Comment by thc202** [comment]
Both `ascanrules` and `ascanrulesAlpha` need to depend on the newer version of `commonlib` (i.e. `>= 1.15.0`).
https://github.com/zaproxy/zap-extensions/blob/22cd4abfd257e1c7d13d5797b7a94d2fd6baf8c8/addOns/ascanrules/ascanrules.gradle.kts#L16-L17

File: addOns/ascanrules/CHANGELOG.md
Comment (thc202) [inline addOns/ascanrules/CHANGELOG.md:13]: - Note that the entries should be added under the `Unreleased` version. This one is old/released already.
- Use `Depend on newer version of Common Library add-on.` instead.

File: addOns/ascanrulesAlpha/CHANGELOG.md
Comment (thc202) [inline addOns/ascanrulesAlpha/CHANGELOG.md:9]:
```suggestion
- Depend on newer version of Common Library add-on.

### Fixed
- Fix time-based false positives in NoSQL Injection - MongoDB scan rule.
```

File: addOns/commonlib/CHANGELOG.md
Comment (thc202) [inline addOns/commonlib/CHANGELOG.md:21]: This one needs to be moved to "Unreleased", can also be something like `Add utilities for time-based checks.`

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/OnlineSimpleLinearRegression.java
Comment (thc202) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/OnlineSimpleLinearRegression.java:22]: Add the `@since` JavaDoc tag to the moved non-test classes, e.g. `@since 1.15.0`.

**Comment by kingthorin** [comment]
It would be nice if we could get https://github.com/zaproxy/zap-extensions/pull/4316 over the finish line then this rebased on top of it.

**Comment by thc202** [comment]
Better the other way around, the changes in that PR are not correct and this PR addresses that (e.g. by not relying on previous requests sent).

**Comment by kingthorin** [comment]
Oh okay! :grinning:

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:366]: I know the problem was there before but the double space may as well be fixed :wink:

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:355]: This along with all other unused code should be removed.

**Comment by itsdivyanshjain** [comment]
@thc202 @kingthorin added suggested changes

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (thc202) [reply addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java]: This is no fully addressed (at least the following are not used: `SLEEP_SHORT_TIME`, `SLEEP_LONG_TIME`, `INSERT_SHORT_TIME`, `INSERT_LONG_TIME`, `aveRtt`, `getAveRtts(List)`, `rtt`, `hadTimeout`).

**Comment by thc202** [comment]
The build is failing, needs a `spotlessApply`.

**Comment by kingthorin** [comment]
The code currently tries to set variables that weren't declared.

**Comment by itsdivyanshjain** [comment]
@thc202 @kingthorin any  update

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:341]: Shouldn't this be inside the loop? Otherwise it will always be the same.

Comment (thc202) [reply addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:341]: Maybe worth adding a test that verifies the payloads are different.

**Comment by kingthorin** [comment]
@itsdivyanshjain are you able to address the latest feedback from @thc202?

**Comment by thc202** [comment]
This is on their roadmap.

**Comment by itsdivyanshjain** [comment]
@kingthorin @thc202 implemented required changes, with additional improvements

**Comment by thc202** [comment]
Build is failing (needs another `spotlessApply`).

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:372]: `TimeUnit.MILLISECONDS.toSeconds(timedMsg.getTimeElapsedMillis())`

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:398]: I'm fine with the found it comment (I guess) but the logging comment probably isn't needed, LOGGER.debug makes it pretty clear.

**Comment by itsdivyanshjain** [comment]
any update?

**Comment by kingthorin** [comment]
Yup, there's outstanding review issues to be addressed.

**Comment by itsdivyanshjain** [comment]
> Yup, there's outstanding review issues to be addressed.

didn't get it. can you elaborate

**Comment by kingthorin** [comment]
🤷‍♂️ it was two weeks ago.

If you’ve addressed all the review comments thus far then it’s just waiting on another review I guess.

**Comment by itsdivyanshjain** [comment]
> 🤷‍♂️ it was two weeks ago.
> 
> If you’ve addressed all the review comments thus far then it’s just waiting on another review I guess.

yeah, i already addressed all the comments since I asked the update.

**Comment by psiinon** [comment]
We rely on contributors telling us that as we dont know if a push addresses any or all of the issues raised 😁

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:73]: If it's no longer a constant it shouldn't be full caps and it should be moved down to 114-124 so that it's grouped with similar vars.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java:294]: Only constants should use FULL CAPS.

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:115]: This isn't a constant, it shouldn't be FULL CAPS.

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:209]: Is this really meant to log the same value twice?

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:204]: If they're changed after substantiation they're no longer CONSTANTS.

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java
Comment (kingthorin) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java:61]: This isn't a CONSTANT.

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java
Comment (kingthorin) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java:86]: I'm not saying this is wrong but maybe it needs a code comment or something? It caught my eye that delay is altered by number of requests 🤷‍♂️ I'm headed for a meeting and don't have time to dig 😜

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (kingthorin) [reply addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:209]: Okay

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java:218]: Remove the comments as well?

Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java:218]: Or relocate them.

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (kingthorin) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java]: Remove or relocate the comments?

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java
Comment (kingthorin) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java:92]: `<=` ?

Would we ever want it to actually be zero?

Comment (itsdivyanshjain) [reply addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java:92]: this would work fine. But, i agree in actual we don't need it to zero

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (itsdivyanshjain) [reply addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java]: removing it because timing utils contains doc string which is sufficient enough to get to know it's working

**Comment by itsdivyanshjain** [comment]
@kingthorin @thc202 looking forward to get this merged if all changes are fine.

File: addOns/ascanrules/CHANGELOG.md
Comment (kingthorin) [inline addOns/ascanrules/CHANGELOG.md:8]:
```suggestion
- Depend on newer version of Common Library add-on.
    - Re-work Command Injection time based payloads/functionality to reduce false positives.
```

File: addOns/commonlib/CHANGELOG.md
Comment (kingthorin) [inline addOns/commonlib/CHANGELOG.md:16]: Add a period please.

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java
Comment (kingthorin) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java:27]: This PR should be rebased and adjusted accordingly. commonlib 1.18 was released last week.

The CHANGELOGs in the ascan rules add-ons likely need to be re-jigged due to releases as well.

**Review by kingthorin — COMMENTED** [review]
I think this is the last of it for me. Thanks for staying on this and pushing us, it really is appreciated. (Both the code and the persistence).

**Review by kingthorin — COMMENTED** [review]
Should be in the Unreleased section

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/OnlineSimpleLinearRegression.java
Comment (kingthorin) [inline addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/OnlineSimpleLinearRegression.java:36]: Sorry I wasn't clear. 1.18.0 is out so this will be upcoming 1.19.0.

**Comment by itsdivyanshjain** [comment]
already done :)

File: addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/OnlineSimpleLinearRegression.java
Comment (itsdivyanshjain) [reply addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/OnlineSimpleLinearRegression.java:36]: done

**Review by kingthorin — APPROVED** [review]
LGTM

**Comment by itsdivyanshjain** [comment]
@thc202 any update

**Comment by kingthorin** [comment]
@KNOXDEV I believe you added the Linear Regression component, do you have any feedback on this?

**Comment by KNOXDEV** [comment]
> @KNOXDEV I believe you added the Linear Regression component, do you have any feedback on this?

This makes sense to me. Good sanity check. Also cool to see this is getting worked into other modules. Thanks for looping me in as always.

**Comment by thc202** [comment]
@itsdivyanshjain could you integrate the latest changes? (We can also take a look at that.)

**Comment by itsdivyanshjain** [comment]
> @itsdivyanshjain could you integrate the latest changes? (We can also take a look at that.)

sure

**Comment by itsdivyanshjain** [comment]
@thc202 @kingthorin , ready for review again

**Comment by thc202** [comment]
I didn't check the latest changes but the build is failing (needs `spottlesApply`).

**Comment by itsdivyanshjain** [comment]
> I didn't check the latest changes but the build is failing (needs `spotlessApply`).

done

**Comment by thc202** [comment]
Could you extract a PR moving just the timing utils to commonlib? That would make things simpler and allow to unblock other PRs.

File: addOns/ascanrules/CHANGELOG.md
Comment (thc202) [inline addOns/ascanrules/CHANGELOG.md:9]: This is covered with the following fixed entry.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java:291]: This rename is not really needed.

File: addOns/commonlib/CHANGELOG.md
Comment (thc202) [inline addOns/commonlib/CHANGELOG.md:9]: Add new line after.

File: addOns/commonlib/src/test/timing/TimingUtilsUnitTest.java
Comment (thc202) [inline addOns/commonlib/src/test/timing/TimingUtilsUnitTest.java:149]: Restore newline.

**Comment by itsdivyanshjain** [comment]
@thc202 New changes are implemented & tested. Do review this again

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:70]: Do these need to be different from the other scan rule? (The name of the first constant could also exactly match the other.)

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:110]: `timeSleepSeconds` to be consistent with the other rule.

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:334]: Delay declaration to line 348.

File: addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java
Comment (thc202) [inline addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:342]: Do after the stop check.

File: addOns/ascanrulesAlpha/src/test/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRuleUnitTest.java
Comment (thc202) [inline addOns/ascanrulesAlpha/src/test/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRuleUnitTest.java:121]: `// Ignore` instead, fail here will not fail the test, just the request (in any case the thread will not be interrupted normally).

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**
